File: mobile/lib/models/medication.dart
/// Medication and DoseLog models.

class Medication {
  final String id;
  final String userId;
  final String name;
  final String category;
  final String form;
  final String? doseAmount;
  final String? doseUnit;
  final String? mealSlot;
  final int? exactHour;
  final int? exactMinute;
  final String frequency;
  final List<int>? freqCustomDays;
  final int? freqWeeklyDay;
  final int? freqMonthlyDay;
  final int? freqHourlyInterval;
  final int? freqHourlyFrom;
  final int? freqHourlyTo;
  final String? startDate;
  final String? endDate;
  final String? bodyArea;
  final int injectionSiteIndex;
  final int stockQuantity;
  final String? stockUnit;
  final int stockAlertThreshold;
  final String color;
  final bool isPrivate;
  final String? notes;
  final bool isActive;
  final String createdBy;
  final DateTime createdAt;

  Medication({
    required this.id,
    required this.userId,
    required this.name,
    this.category = 'medicine',
    this.form = 'tablet',
    this.doseAmount,
    this.doseUnit,
    this.mealSlot,
    this.exactHour,
    this.exactMinute,
    this.frequency = 'daily',
    this.freqCustomDays,
    this.freqWeeklyDay,
    this.freqMonthlyDay,
    this.freqHourlyInterval,
    this.freqHourlyFrom,
    this.freqHourlyTo,
    this.startDate,
    this.endDate,
    this.bodyArea,
    this.injectionSiteIndex = 0,
    this.stockQuantity = 0,
    this.stockUnit,
    this.stockAlertThreshold = 5,
    this.color = '#059669',
    this.isPrivate = true,
    this.notes,
    this.isActive = true,
    required this.createdBy,
    required this.createdAt,
  });

  factory Medication.fromJson(Map<String, dynamic> json) {
    return Medication(
      id: json['id'],
      userId: json['user_id'],
      name: json['name'],
      category: json['category'] ?? 'medicine',
      form: json['form'] ?? 'tablet',
      doseAmount: json['dose_amount'],
      doseUnit: json['dose_unit'],
      mealSlot: json['meal_slot'],
      exactHour: json['exact_hour'],
      exactMinute: json['exact_minute'],
      frequency: json['frequency'] ?? 'daily',
      freqCustomDays: json['freq_custom_days'] != null
          ? List<int>.from(json['freq_custom_days'])
          : null,
      freqWeeklyDay: json['freq_weekly_day'],
      freqMonthlyDay: json['freq_monthly_day'],
      freqHourlyInterval: json['freq_hourly_interval'],
      freqHourlyFrom: json['freq_hourly_from'],
      freqHourlyTo: json['freq_hourly_to'],
      startDate: json['start_date'],
      endDate: json['end_date'],
      bodyArea: json['body_area'],
      injectionSiteIndex: json['injection_site_index'] ?? 0,
      stockQuantity: json['stock_quantity'] ?? 0,
      stockUnit: json['stock_unit'],
      stockAlertThreshold: json['stock_alert_threshold'] ?? 5,
      color: json['color'] ?? '#059669',
      isPrivate: json['is_private'] ?? true,
      notes: json['notes'],
      isActive: json['is_active'] ?? true,
      createdBy: json['created_by'],
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  /// Scheduled time as "8:30 AM" string.
  String get scheduledTimeStr {
    if (exactHour == null) return '';
    final h = exactHour! > 12 ? exactHour! - 12 : (exactHour == 0 ? 12 : exactHour!);
    final ampm = exactHour! >= 12 ? 'PM' : 'AM';
    final m = (exactMinute ?? 0).toString().padLeft(2, '0');
    return '$h:$m $ampm';
  }

  /// Dose display string like "1 tablet" or "10 ml".
  String get doseDisplay {
    if (doseAmount == null && doseUnit == null) return '';
    return '${doseAmount ?? ''} ${doseUnit ?? ''}'.trim();
  }

  /// Whether stock is low.
  bool get isLowStock => stockQuantity > 0 && stockQuantity <= stockAlertThreshold;

  /// Parse hex color to Color int.
  int get colorValue {
    final hex = color.replaceAll('#', '');
    return int.parse('FF$hex', radix: 16);
  }
}


class DoseLog {
  final String id;
  final String medicationId;
  final String userId;
  final String scheduledDate;
  final String? scheduledTime;
  final String status; // taken, skipped, missed, snoozed
  final DateTime? actualTime;
  final String loggedBy;
  final String? injectionSite;
  final String? notes;
  final DateTime createdAt;

  DoseLog({
    required this.id,
    required this.medicationId,
    required this.userId,
    required this.scheduledDate,
    this.scheduledTime,
    required this.status,
    this.actualTime,
    required this.loggedBy,
    this.injectionSite,
    this.notes,
    required this.createdAt,
  });

  factory DoseLog.fromJson(Map<String, dynamic> json) {
    return DoseLog(
      id: json['id'],
      medicationId: json['medication_id'],
      userId: json['user_id'],
      scheduledDate: json['scheduled_date'],
      scheduledTime: json['scheduled_time'],
      status: json['status'],
      actualTime: json['actual_time'] != null
          ? DateTime.parse(json['actual_time'])
          : null,
      loggedBy: json['logged_by'],
      injectionSite: json['injection_site'],
      notes: json['notes'],
      createdAt: DateTime.parse(json['created_at']),
    );
  }
}


class TodayScheduleItem {
  final Medication medication;
  final DoseLog? doseLog;
  final bool isDue;
  final bool isMissed;

  TodayScheduleItem({
    required this.medication,
    this.doseLog,
    this.isDue = false,
    this.isMissed = false,
  });

  factory TodayScheduleItem.fromJson(Map<String, dynamic> json) {
    return TodayScheduleItem(
      medication: Medication.fromJson(json['medication']),
      doseLog: json['dose_log'] != null
          ? DoseLog.fromJson(json['dose_log'])
          : null,
      isDue: json['is_due'] ?? false,
      isMissed: json['is_missed'] ?? false,
    );
  }

  String get status {
    if (doseLog != null) return doseLog!.status;
    if (isMissed) return 'missed';
    if (isDue) return 'due';
    return 'pending';
  }
}

File: backend/app/schemas/medication.py
"""Medication & DoseLog schemas."""

from pydantic import BaseModel, Field
from typing import Optional, List
from uuid import UUID
from datetime import date, datetime, time


class MedicationCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    category: str = "medicine"
    form: str = "tablet"
    dose_amount: Optional[str] = None
    dose_unit: Optional[str] = None
    meal_slot: Optional[str] = None
    exact_hour: Optional[int] = Field(None, ge=0, le=23)
    exact_minute: Optional[int] = Field(None, ge=0, le=59)
    frequency: str = "daily"
    freq_custom_days: Optional[List[int]] = None
    freq_weekly_day: Optional[int] = Field(None, ge=0, le=6)
    freq_monthly_day: Optional[int] = Field(None, ge=1, le=28)
    freq_hourly_interval: Optional[int] = Field(None, ge=1, le=24)
    freq_hourly_from: Optional[int] = Field(None, ge=0, le=23)
    freq_hourly_to: Optional[int] = Field(None, ge=0, le=23)
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    body_area: Optional[str] = None
    stock_quantity: int = 0
    stock_unit: Optional[str] = None
    stock_alert_threshold: int = 5
    color: str = "#059669"
    is_private: bool = True
    notes: Optional[str] = None


class MedicationUpdate(BaseModel):
    name: Optional[str] = None
    category: Optional[str] = None
    form: Optional[str] = None
    dose_amount: Optional[str] = None
    dose_unit: Optional[str] = None
    meal_slot: Optional[str] = None
    exact_hour: Optional[int] = None
    exact_minute: Optional[int] = None
    frequency: Optional[str] = None
    freq_custom_days: Optional[List[int]] = None
    freq_weekly_day: Optional[int] = None
    freq_monthly_day: Optional[int] = None
    freq_hourly_interval: Optional[int] = None
    freq_hourly_from: Optional[int] = None
    freq_hourly_to: Optional[int] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    body_area: Optional[str] = None
    stock_quantity: Optional[int] = None
    stock_unit: Optional[str] = None
    stock_alert_threshold: Optional[int] = None
    color: Optional[str] = None
    is_private: Optional[bool] = None
    notes: Optional[str] = None
    is_active: Optional[bool] = None


class MedicationOut(BaseModel):
    id: UUID
    user_id: UUID
    name: str
    category: str
    form: str
    dose_amount: Optional[str] = None
    dose_unit: Optional[str] = None
    meal_slot: Optional[str] = None
    exact_hour: Optional[int] = None
    exact_minute: Optional[int] = None
    frequency: str
    freq_custom_days: Optional[List[int]] = None
    freq_weekly_day: Optional[int] = None
    freq_monthly_day: Optional[int] = None
    freq_hourly_interval: Optional[int] = None
    freq_hourly_from: Optional[int] = None
    freq_hourly_to: Optional[int] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    body_area: Optional[str] = None
    injection_site_index: int = 0
    stock_quantity: int = 0
    stock_unit: Optional[str] = None
    stock_alert_threshold: int = 5
    color: str
    is_private: bool
    notes: Optional[str] = None
    is_active: bool
    created_by: UUID
    created_at: datetime

    model_config = {"from_attributes": True}


class DoseLogCreate(BaseModel):
    status: str = "taken"  # taken, skipped, snoozed
    notes: Optional[str] = None


class DoseLogOut(BaseModel):
    id: UUID
    medication_id: UUID
    user_id: UUID
    scheduled_date: date
    scheduled_time: Optional[time] = None
    status: str
    actual_time: Optional[datetime] = None
    logged_by: UUID
    injection_site: Optional[str] = None
    notes: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class TodayScheduleItem(BaseModel):
    medication: MedicationOut
    dose_log: Optional[DoseLogOut] = None
    is_due: bool = False
    is_missed: bool = False
